lib/src/net/endpoint.dart

```diff
@@ -104,36 +102,30 @@ class Endpoint implements Outbox {
     _coapStack.sendEmptyMessage(exchange, message);
   }
 
-  void _receiveData(final CoapDataReceivedEvent event) {
-    // clone the data, in case other objects want to do stuff with it, too
-    final data = Uint8Buffer()..addAll(event.data);
-    // Return if we have no data, should not happen but be defensive
-    final decoder = config.spec.newMessageDecoder(data);
-    if (decoder.isRequest) {
-      CoapRequest? request;
-      try {
-        request = decoder.decodeRequest();
-      } on Exception catch (_) {
-        if (!decoder.isReply) {
-          // Manually build RST from raw information
-          final rst = CoapEmptyMessage(CoapMessageType.rst)
-            ..destination = event.address
-            ..id = decoder.id;
-          _eventBus.fire(CoapSendingEmptyMessageEvent(rst));
-          _socket.send(_serializeEmpty(rst), rst.destination);
-        }
-        return;
-      }
+  void _receiveMessage(final CoapMessageReceivedEvent event) {
+    final message = event.coapMessage;
 
-      request!.source = event.address;
+    if (message == null) {
+      return;
+    }
+
+    if (message.needsRejection) {
+      _reject(message);
+      return;
+    }
```

Comment:
@shamblett I have managed to find out that the ping issue is caused by this check. I will open a follow-up PR shortly.